posthog/hogql_queries/insights/retention_query_runner.py

```diff
@@ -0,0 +1,316 @@
+from datetime import datetime, timedelta
+from math import ceil
+from typing import Any, Dict
+from typing import Optional
+
+from posthog.caching.insights_api import BASE_MINIMUM_INSIGHT_REFRESH_INTERVAL, REDUCED_MINIMUM_INSIGHT_REFRESH_INTERVAL
+from posthog.caching.utils import is_stale
+from posthog.constants import (
+    TREND_FILTER_TYPE_EVENTS,
+    RetentionQueryType,
+)
+from posthog.hogql import ast
+from posthog.hogql.parser import parse_select
+from posthog.hogql.printer import to_printed_hogql
+from posthog.hogql.property import property_to_expr, entity_to_expr
+from posthog.hogql.query import execute_hogql_query
+from posthog.hogql.timings import HogQLTimings
+from posthog.hogql_queries.query_runner import QueryRunner
+from posthog.hogql_queries.utils.query_date_range import QueryDateRangeWithIntervals
+from posthog.models import Team
+from posthog.models.filters.mixins.utils import cached_property
+from posthog.queries.util import correct_result_for_sampling
+from posthog.schema import (
+    HogQLQueryModifiers,
+    RetentionQueryResponse,
+    IntervalType,
+    RetentionFilter,
+)
+from posthog.schema import RetentionQuery, RetentionType
+
+DEFAULT_INTERVAL = IntervalType("day")
+DEFAULT_TOTAL_INTERVALS = 11
+
+
+class RetentionQueryRunner(QueryRunner):
+    query: RetentionQuery
+    query_type = RetentionQuery
+
+    def __init__(
+        self,
+        query: RetentionQuery | Dict[str, Any],
+        team: Team,
+        timings: Optional[HogQLTimings] = None,
+        modifiers: Optional[HogQLQueryModifiers] = None,
+        in_export_context: Optional[bool] = None,
+    ):
+        super().__init__(query, team=team, timings=timings, modifiers=modifiers, in_export_context=in_export_context)
+        if not self.query.retentionFilter:
+            self.query.retentionFilter = RetentionFilter()  # TODO: Clarify default filter
+
+        self.target_entity = self.query.retentionFilter.target_entity or {
+            "id": "$pageview",
+            "type": TREND_FILTER_TYPE_EVENTS,
+        }
```

Comment:
I know you didn't modify `renteintionFilter`, but it's a bit weird to see a dict, when everything else operates on dataclasses. Now that you know what shape the data takes here, can you add the correct types into the old retention filter type:
https://github.com/PostHog/posthog/blob/55c7f08d61f6378cddb1162353adb1677c3d4e9d/frontend/src/types.ts#L1882-L1883

Reply:
You mean to use also pydantic dataclasses here? I also thought that would be useful and I think I checked and it trickles down quite a lot... would be additional work, but can be done 👍🏻